src/thor/triplegbuilder.cc

```diff
@@ -289,6 +307,25 @@ void SetShapeAttributes(const AttributesController& controller,
       distance *= coef;
       shift = 1;
     }
+    if (controller.attributes.at(kShapeAttributesClosure)) {
+      // Process closure annotations
+      if (cut_itr->closed) {
+        // Found a closure. Fetch a new annotation, or the last closure
+        // annotation if it does not have an end index set (meaning the shape
+        // is still within an existing closure)
+        ::valhalla::TripLeg_Closure* closure = fetch_or_create_closure_annotation(leg);
+        if (!closure->has_begin_shape_index()) {
+          closure->set_begin_shape_index(i - 1);
```

Comment:
How come it's `i - 1` and not just `i`?

Reply:
Since this loops [starts from the 2nd shape index](https://github.com/valhalla/valhalla/pull/2816/files#diff-fd61a9dafa6119b300d3fcc49f59456388c00d6851fb8cc325141be44daf58c4R295). It does that since we create cuts only if speed changes in subsequent shape segments.